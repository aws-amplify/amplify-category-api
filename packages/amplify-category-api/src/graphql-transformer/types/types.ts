export type ConstructResourceMeta = {
  rootStack?: StackMeta;
  nestedStack?: StackMeta;
  resourceName: string;
  resourceType: string;
};

export type StackMeta = {
  stackName: string;
  stackType: string;
};
