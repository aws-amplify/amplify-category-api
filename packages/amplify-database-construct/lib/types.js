"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidHlwZXMuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi9zcmMvdHlwZXMudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IiIsInNvdXJjZXNDb250ZW50IjpbIi8qKlxuICogSW5wdXQgcHJvcHMgZm9yIHRoZSBBbXBsaWZ5RGF0YWJhc2UgY29uc3RydWN0LlxuICovXG5leHBvcnQgaW50ZXJmYWNlIEFtcGxpZnlEYXRhYmFzZVByb3BzIHtcbiAgcmVhZG9ubHkgZGVmaW5pdGlvbjogc3RyaW5nO1xufVxuIl19