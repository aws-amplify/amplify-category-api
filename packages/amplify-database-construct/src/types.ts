import { IVpc } from 'aws-cdk-lib/aws-ec2';
import { IDatabaseCluster } from 'aws-cdk-lib/aws-rds';

/**
 * Input props for the AmplifyDatabase construct.
 */
export interface AmplifyDatabaseProps {
  readonly vpc: IVpc;
}

export interface AmplifyDatabaseResources {
  /**
   * The database cluster created by the construct.
   */
  readonly databaseCluster: IDatabaseCluster;
}
