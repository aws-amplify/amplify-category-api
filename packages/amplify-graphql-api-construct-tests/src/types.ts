/**
 * Represents stack outputs generated by Amplify Auth Construct.
 */
export type AuthConstructStackOutputs = {
  authRegion: string;
  userPoolId: string;
  webClientId: string;
  identityPoolId: string;
};
