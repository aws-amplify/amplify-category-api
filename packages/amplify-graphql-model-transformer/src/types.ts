import { SyncConfig } from '@aws-amplify/graphql-transformer-core';

export type ModelTransformerOptions = {
  EnableDeletionProtection?: boolean;
  SyncConfig?: SyncConfig;
};
