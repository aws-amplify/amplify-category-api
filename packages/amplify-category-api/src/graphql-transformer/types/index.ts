export { ConstructResourceMeta, StackMeta } from './types';
export { stacksTypes, getStackMeta, convertToAppsyncResourceObj } from './utils';
