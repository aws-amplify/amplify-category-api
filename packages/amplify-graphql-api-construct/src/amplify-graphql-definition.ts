import * as os from 'os';
import { SchemaFile } from 'aws-cdk-lib/aws-appsync';
import { isSqlStrategy } from '@aws-amplify/graphql-transformer-core';
import { IAmplifyGraphqlDefinition } from './types';
import { constructDataSourceStrategies } from './internal';
import { CustomSqlDataSourceStrategy, ModelDataSourceStrategy } from './model-datasource-strategy-types';
import {
  constructCustomSqlDataSourceStrategies,
  schemaByMergingDefinitions,
  validateDataSourceStrategy,
} from './internal/data-source-config';

export const DEFAULT_MODEL_DATA_SOURCE_STRATEGY: ModelDataSourceStrategy = {
  dbType: 'DYNAMODB',
  provisionStrategy: 'DEFAULT',
};

/**
 * Class exposing utilities to produce IAmplifyGraphqlDefinition objects given various inputs.
 */
export class AmplifyGraphqlDefinition {
  /**
   * Produce a schema definition from a string input.
   * @param schema the graphql input as a string
   * @param dataSourceStrategy the provisioning definition for datasources that resolve `@model`s and custom SQL statements in this schema.
   * The DynamoDB from CloudFormation will be used by default.
   * @returns a fully formed amplify graphql definition
   */
  static fromString(
    schema: string,
    dataSourceStrategy: ModelDataSourceStrategy = DEFAULT_MODEL_DATA_SOURCE_STRATEGY,
  ): IAmplifyGraphqlDefinition {
    validateDataSourceStrategy(dataSourceStrategy);
    return {
      schema,
      functionSlots: [],
      referencedLambdaFunctions: {},
      dataSourceStrategies: constructDataSourceStrategies(schema, dataSourceStrategy),
      customSqlDataSourceStrategies: constructCustomSqlDataSourceStrategies(schema, dataSourceStrategy),
    };
  }

  /**
   * Convert one or more appsync SchemaFile objects into an Amplify Graphql Schema, binding them to a DynamoDB data source.
   * @param filePaths one or more paths to the graphql files to process
   * @returns a fully formed amplify graphql definition, whose models will be resolved by DynamoDB tables created during deployment.
   */
  static fromFiles(...filePaths: string[]): IAmplifyGraphqlDefinition {
    if (!Array.isArray(filePaths)) {
      filePaths = [filePaths];
    }
    const schema = filePaths.map((filePath) => new SchemaFile({ filePath }).definition).join(os.EOL);
    return AmplifyGraphqlDefinition.fromString(schema, DEFAULT_MODEL_DATA_SOURCE_STRATEGY);
  }

  /**
   * Convert one or more appsync SchemaFile objects into an Amplify Graphql Schema
   * @param filePaths one or more paths to the graphql files to process
   * @param dataSourceStrategy the provisioning definition for datasources that resolve `@model`s in this schema. The DynamoDB from
   * CloudFormation will be used by default.
   * @returns a fully formed amplify graphql definition
   */
  static fromFilesAndStrategy(
    filePaths: string | string[],
    dataSourceStrategy: ModelDataSourceStrategy = DEFAULT_MODEL_DATA_SOURCE_STRATEGY,
  ): IAmplifyGraphqlDefinition {
    if (!Array.isArray(filePaths)) {
      filePaths = [filePaths];
    }
    const schema = filePaths.map((filePath) => new SchemaFile({ filePath }).definition).join(os.EOL);
    return AmplifyGraphqlDefinition.fromString(schema, dataSourceStrategy);
  }

  /**
   * Combines multiple IAmplifyGraphqlDefinitions into a single definition.
   * @param definitions the definitions to combine
   */
  static combine(definitions: IAmplifyGraphqlDefinition[]): IAmplifyGraphqlDefinition {
    if (definitions.length === 0) {
      throw new Error('The definitions of amplify GraphQL cannot be empty.');
    }
    if (definitions.length === 1) {
      return definitions[0];
    }

    // A strategy will be present multiple times in a given definition: once per model. We'll create a unique list per definition to ensure
    // no reuse across definitions.
    let combinedStrategyNames: string[] = [];
    for (const definition of definitions) {
      const definitionStrategyNames = new Set<string>();
      for (const strategy of Object.values(definition.dataSourceStrategies)) {
        if (!isSqlStrategy(strategy)) {
          continue;
        }
        const strategyName = strategy.name;
        if (combinedStrategyNames.includes(strategyName)) {
          throw new Error(
            `The SQL-based ModelDataSourceStrategy '${strategyName}' was found in multiple definitions, but a strategy name cannot be ` +
              "shared between definitions. To specify a SQL-based API with schemas across multiple files, use 'fromFilesAndStrategy'",
          );
        }
        definitionStrategyNames.add(strategyName);
      }
      combinedStrategyNames = [...combinedStrategyNames, ...definitionStrategyNames];
    }

    const customSqlDataSourceStrategies = definitions.reduce(
      (acc, cur) => [...acc, ...(cur.customSqlDataSourceStrategies ?? [])],
      [] as CustomSqlDataSourceStrategy[],
    );

    const mergedSchema = schemaByMergingDefinitions(definitions);

    return {
      schema: mergedSchema,
      functionSlots: [],
      referencedLambdaFunctions: definitions.reduce((acc, cur) => ({ ...acc, ...cur.referencedLambdaFunctions }), {}),
      dataSourceStrategies: definitions.reduce((acc, cur) => ({ ...acc, ...cur.dataSourceStrategies }), {}),
      customSqlDataSourceStrategies,
    };
  }
}
