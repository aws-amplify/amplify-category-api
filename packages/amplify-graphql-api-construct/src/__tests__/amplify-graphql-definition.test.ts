import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { AmplifyGraphqlDefinition } from '../amplify-graphql-definition';
import { ModelDataSourceBinding, SqlModelDataSourceBinding } from '../types';

const TEST_SCHEMA = /* GraphQL */ `
  type Todo @model {
    id: ID!
    content: String!
  }
`;

describe('AmplifyGraphqlDefinition', () => {
  describe('fromString', () => {
    it('returns the provided string and no functions', () => {
      const definition = AmplifyGraphqlDefinition.fromString(TEST_SCHEMA);
      expect(definition.schema).toEqual(TEST_SCHEMA);
      expect(definition.functionSlots.length).toEqual(0);
    });
  });

  describe('fromFiles', () => {
    let tmpDir: string;

    beforeEach(() => {
      tmpDir = fs.mkdtempSync('fromFiles');
    });

    afterEach(() => {
      if (fs.existsSync(tmpDir)) fs.rmdirSync(tmpDir, { recursive: true });
    });

    it('extracts the definition from a single schema file', () => {
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      const definition = AmplifyGraphqlDefinition.fromFiles(schemaFilePath);
      expect(definition.schema).toEqual(TEST_SCHEMA);
      expect(definition.functionSlots.length).toEqual(0);
    });

    it('extracts the definition from the schema files, appended in-order', () => {
      const rdsTestSchema = /* GraphQL */ `
        type Blog @model {
          id: ID!
          posts: [Post] @hasMany
        }

        type Post @model {
          id: ID!
          blog: Blog @belongsTo
        }
      `;
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      const rdsSchemaFilePath = path.join(tmpDir, 'schema.rds.graphql');
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      fs.writeFileSync(rdsSchemaFilePath, rdsTestSchema);
      const definition = AmplifyGraphqlDefinition.fromFiles(schemaFilePath, rdsSchemaFilePath);
      expect(definition.schema).toEqual(`${TEST_SCHEMA}${os.EOL}${rdsTestSchema}`);
      expect(definition.functionSlots.length).toEqual(0);
    });

    it('binds to a DynamoDB data source', () => {
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      const definition = AmplifyGraphqlDefinition.fromFiles(schemaFilePath);
      expect(definition.modelDataSourceBinding).toEqual({ bindingType: 'DynamoDB' });
    });
  });

  describe('fromFilesAndBinding', () => {
    let tmpDir: string;

    const defaultBinding: ModelDataSourceBinding = {
      bindingType: 'DynamoDB',
    };

    beforeEach(() => {
      tmpDir = fs.mkdtempSync('fromFiles');
    });

    afterEach(() => {
      if (fs.existsSync(tmpDir)) fs.rmdirSync(tmpDir, { recursive: true });
    });

    it('extracts the definition from a single schema file', () => {
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      const definition = AmplifyGraphqlDefinition.fromFilesAndBinding([schemaFilePath], defaultBinding);
      expect(definition.schema).toEqual(TEST_SCHEMA);
      expect(definition.functionSlots.length).toEqual(0);
    });

    it('extracts the definition from the schema files, appended in-order', () => {
      const rdsTestSchema = /* GraphQL */ `
        type Blog @model {
          id: ID!
          posts: [Post] @hasMany
        }

        type Post @model {
          id: ID!
          blog: Blog @belongsTo
        }
      `;
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      const rdsSchemaFilePath = path.join(tmpDir, 'schema.rds.graphql');
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      fs.writeFileSync(rdsSchemaFilePath, rdsTestSchema);
      const definition = AmplifyGraphqlDefinition.fromFilesAndBinding([schemaFilePath, rdsSchemaFilePath], defaultBinding);
      expect(definition.schema).toEqual(`${TEST_SCHEMA}${os.EOL}${rdsTestSchema}`);
      expect(definition.functionSlots.length).toEqual(0);
    });

    it('binds to a dynamo data source', () => {
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      const definition = AmplifyGraphqlDefinition.fromFilesAndBinding([schemaFilePath], defaultBinding);
      expect(definition.modelDataSourceBinding).toEqual(defaultBinding);
    });

    it('binds to a sql data source', () => {
      const schemaFilePath = path.join(tmpDir, 'schema.graphql');
      const binding: SqlModelDataSourceBinding = {
        bindingType: 'MySQL',
        vpcConfiguration: {
          vpcId: 'vpc-1234abcd',
          securityGroupIds: ['sg-123'],
          subnetAvailabilityZones: [{ subnetId: 'subnet-123', availabilityZone: 'us-east-1a' }],
        },
        dbConnectionConfig: {
          hostnameSsmPath: '/ssm/path/hostnameSsmPath',
          portSsmPath: '/ssm/path/portSsmPath',
          usernameSsmPath: '/ssm/path/usernameSsmPath',
          passwordSsmPath: '/ssm/path/passwordSsmPath',
          databaseNameSsmPath: '/ssm/path/databaseNameSsmPath',
        },
      };
      fs.writeFileSync(schemaFilePath, TEST_SCHEMA);
      const definition = AmplifyGraphqlDefinition.fromFilesAndBinding([schemaFilePath], binding);
      expect(definition.modelDataSourceBinding).toEqual(binding);
    });
  });
});
