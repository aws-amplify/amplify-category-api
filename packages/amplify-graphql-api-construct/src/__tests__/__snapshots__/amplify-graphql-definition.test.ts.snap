// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmplifyGraphqlDefinition combine returns the correct definition after the combination, preserving relationship and auth directives 1`] = `
"type Todo @model {
  id: ID!
  content: String!
}

type Blog @model @auth(rules: [{allow: owner}]) {
  id: ID!
  posts: [Post] @hasMany
}

type Post @model @auth(rules: [{allow: owner}, {allow: public, operations: [read]}]) {
  id: ID!
  blog: Blog @belongsTo
}

type Mutation {
  updateOnlyOwner: [Int] @sql(statement: \\"UPDATE foo SET id = 1; SELECT 1\\") @auth(rules: [{allow: owner}])
  updateAllowPublic: [Int] @sql(statement: \\"UPDATE foo SET id = 1; SELECT 1\\") @auth(rules: [{allow: public}])
}

type Query {
  getOnlyOwner: [Int] @sql(statement: \\"SELECT 1\\") @auth(rules: [{allow: owner}])
  getAllowPublic: [Int] @sql(statement: \\"SELECT 1\\") @auth(rules: [{allow: public}])
}
"
`;
